### PR TITLE
Added rocksdb with transactions using Yiyuanliu branch of rust-rocksdb.

### PR DESCRIPTION
**Pull request by Mo0nbase**
TODO:
- Fix the lifetimes for the raw prefix search
- Change the API and Database trait to support self referencing transactions 
- Use native rocksdb column families instead of prefixing (e.g. collecting signature shares for a given token and deleting the column later)
- Perhaps create inherited traits for basic database -> transactional database -> transactional + column family database
- Crate generic trait for expected "transaction" behaviors and functionality
- Get rid of sled_impl and replace all references
- Remove Batch.rs  
- Improve UT and add coverage for all expected trait functionality
Note: transactions are soon to be merged into rust-rocksdb we are simply waiting on a couple changes for the next rocksdb you can track the issue here https://github.com/rust-rocksdb/rust-rocksdb/pull/565/files#. Obviously we will want to change to master as soon as possible.

### CODE REVIEW

**Comment by Maan2003** [comment]
I think most of these TODOs should be addressed in separate PRs.

For lifetimes, feel free to add 98eae77d2b99ec1afa22c8fd6006a81b41bc7656 to the PR

**Comment by Maan2003** [comment]
On wasm side, I suggest making rocksdb optional dependency, and enabling it in `minimint`, `ln-gateway` and `mint-client-cli`

**Comment by Mo0nbase** [comment]
Okay the build should be fixed although with the inclusion of C code GitHub wont be able to compile and test without clang tools (I'm not sure how that works). As for the TODO's I think I'll limit this PR to adding the remaining tests for the existing database trait and making rocksdb an optional dependency and enabling it in those files Maan2003 mentioned. The other changes such as the migration to transactions and the other TODO's will have separate PR's.

**Comment by Maan2003** [comment]
notes for CI:
- run cargo fmt
- install clang in nix
- make rocksdb optional dependency, so it is not used on wasm

**Comment by Maan2003** [comment]
feel free to include 9ebfa353051a5e5904a58f9e94fc2888507d954f for wasm fixes

File: minimint-api/Cargo.toml [hunk 1]
```diff
@@ -15,6 +15,7 @@ http-types = "2.12.0"
 minimint-derive = { path = "../minimint-derive" }
 rand = "0.6.0"
 secp256k1-zkp = { git = "https://github.com/fedimint/rust-secp256k1-zkp", branch = "minimint", features = [ "use-serde", "bitcoin_hashes", "global-context" ] }
+rocksdb = { git = "https://github.com/yiyuanliu/rust-rocksdb", branch = "transaction", optional = true }
```
Comment (justinmoon) [inline minimint-api/Cargo.toml:18]: Let's put our forked dependency in the `[patch.crates-io]` section of the root `Cargo.toml` [like this](https://github.com/fedimint/minimint/blob/52d39f5cdfcb6cd3a0aa9648720d7f3c7e28477b/Cargo.toml#L18).

Let's use our own [fedimint/rust-rocksdb](https://github.com/fedimint/rust-rocksdb/tree/minimint) fork which I just created. The `minimint` branch there should be identical to yiyuanliu's branch.

**Comment by wiredhikari** [comment]
You can try a `nix flake update`

**Comment by elsirion** [comment]
Awesome that the tests finally run, you can probably clean up the git history a bit and squash some commits, then this should be ready for review!

**Comment by justinmoon** [comment]
What's the status here?

**Comment by Maan2003** [comment]
I have rebased and squashed some commits in https://github.com/Maan2003/minimint/commits/db-rebase

**Review by elsirion — APPROVED** [review]
Great work!

**Pull request merged**